module/zfs/spa.c

```diff
@@ -3326,6 +3338,106 @@ spa_start_livelist_condensing_thread(spa_t *spa)
 	    spa_livelist_condense_cb, spa, minclsyspri);
 }
 
+static int
+spa_rrd_write(spa_t *spa, dmu_tx_t *tx, rrd_t *rrd, const char *what,
+    boolean_t insert)
+{
+	static uint8_t buf[sizeof (*rrd)];
+
+	rrd_pack(rrd, buf);
+
+	if (insert) {
+		return (zap_add(spa_meta_objset(spa), DMU_POOL_DIRECTORY_OBJECT,
+		    what, 1, sizeof (*rrd), buf, tx));
+	}
+
+	return (zap_update(spa_meta_objset(spa),
+	    DMU_POOL_DIRECTORY_OBJECT, what, 1, sizeof (*rrd), buf,
+	    tx));
+}
+
+static void
+spa_sync_time_logger(spa_t *spa, uint64_t txg)
+{
+	uint64_t curtime;
+	boolean_t insert;
+	dmu_tx_t *tx;
+
+	if (!spa_writeable(spa)) {
+		return;
+	}
+	if (txg == spa->spa_last_noted_txg) {
+		return;
+	}
+	curtime = gethrestime_sec();
+	if (curtime < spa->spa_last_noted_txg_time + spa_note_txg_time) {
+		return;
+	}
+
+	spa->spa_last_noted_txg_time = curtime;
+	spa->spa_last_noted_txg = txg;
+
+	mutex_enter(&spa->spa_txg_log_time_lock);
+	dbrrd_add(&spa->spa_txg_log_time, curtime, txg);
+
+	if (curtime < spa->spa_last_flush_txg_time + spa_flush_txg_time) {
+		goto out;
+	}
+	spa->spa_last_flush_txg_time = curtime;
+
+	insert = (zap_contains(spa_meta_objset(spa), DMU_POOL_DIRECTORY_OBJECT,
+	    DMU_POOL_TXG_LOG_TIME_MINUTES) == ENOENT);
+
+	tx = dmu_tx_create_assigned(spa_get_dsl(spa), txg);
+	VERIFY0(spa_rrd_write(spa, tx, &spa->spa_txg_log_time.dbr_minutes,
+	    DMU_POOL_TXG_LOG_TIME_MINUTES, insert));
+	VERIFY0(spa_rrd_write(spa, tx, &spa->spa_txg_log_time.dbr_days,
+	    DMU_POOL_TXG_LOG_TIME_DAYS, insert));
+	VERIFY0(spa_rrd_write(spa, tx, &spa->spa_txg_log_time.dbr_months,
+	    DMU_POOL_TXG_LOG_TIME_MONTHS, insert));
+	dmu_tx_commit(tx);
+
+out:
+	mutex_exit(&spa->spa_txg_log_time_lock);
+}
+
+static int
+spa_load_txg_log_time(spa_t *spa)
+{
+	int error;
+
+	error = zap_lookup(spa->spa_meta_objset, DMU_POOL_DIRECTORY_OBJECT,
+	    DMU_POOL_TXG_LOG_TIME_MINUTES,
+	    1, sizeof (spa->spa_txg_log_time.dbr_minutes),
+	    &spa->spa_txg_log_time.dbr_minutes);
+
+	if (error == ENOENT)
+		return (0);
+	if (error != 0)
+		return (error);
+
+	error = zap_lookup(spa->spa_meta_objset, DMU_POOL_DIRECTORY_OBJECT,
+	    DMU_POOL_TXG_LOG_TIME_DAYS,
+	    1, sizeof (spa->spa_txg_log_time.dbr_days),
+	    &spa->spa_txg_log_time.dbr_days);
+	if (error != 0)
+		return (error);
+
+	error = zap_lookup(spa->spa_meta_objset, DMU_POOL_DIRECTORY_OBJECT,
+	    DMU_POOL_TXG_LOG_TIME_MONTHS,
+	    1, sizeof (spa->spa_txg_log_time.dbr_months),
+	    &spa->spa_txg_log_time.dbr_months);
+	if (error != 0)
+		return (error);
+
+	rrd_ntoh(&spa->spa_txg_log_time.dbr_minutes);
+	rrd_ntoh(&spa->spa_txg_log_time.dbr_days);
+	rrd_ntoh(&spa->spa_txg_log_time.dbr_months);
```

Comment:
ZAP has native support for byte swapping if you specify that you store 8-byte integers (same as 2- and 4-byte).  But instead you are storing 8 1-byte integers to create yourself a problem.  Array of 1-byte integers makes sense for opaque data blobs, but numbers should be stores as numbers.  Actually, since all fields in the RRD value are uint64_t, I think you could store it as array of 8-byte integers and ZAP would do all the byte-swapping for you. It would be much cleaner than your manual byte swapping!

Reply:
Thanks done.